fixes interactive.harbor issues. ref #3418

ghostnumber7 opened the pull request:
fiexes #3418

smimram commented:
Thanks for this PR! It would be nice if `escape_html` could be moved to `string.liq` and renamed to `string.escape.html` since it could be useful outside of this context.

ghostnumber7 commented:
> Thanks for this PR! It would be nice if `escape_html` could be moved to `string.liq` and renamed to `string.escape.html` since it could be useful outside of this context.

Moved. I'm not sure about the "docs" entry tho.

vitoyucepi commented on src/libs/string.liq:
```diff
@@ -381,3 +381,21 @@ let string.char.ascii.number = list.init(10, fun (c) -> c + 48)
+      elsif c == "<" then "&lt;"
+      elsif c == ">" then "&gt;"
+      elsif c == "\"" then "&quot;"
+      elsif c == "'" then "&apos;"
```
Is it safe to use `&apos;` instead of `&#39;`?

ghostnumber7 replied:
It should be the same

vitoyucepi commented on src/libs/string.liq:
```diff
@@ -381,3 +381,21 @@ let string.char.ascii.number = list.init(10, fun (c) -> c + 48)
+      if c == "&" then "&amp;"
+      elsif c == "<" then "&lt;"
+      elsif c == ">" then "&gt;"
+      elsif c == "\"" then "&quot;"
```
```suggestion
      elsif c == '"' then "&quot;"
```
I think it's OK to use single quotes to prevent escaping, but the linter may not be happy.

vitoyucepi commented on src/libs/string.liq:
```diff
@@ -381,3 +381,21 @@ let string.char.ascii.number = list.init(10, fun (c) -> c + 48)
+# Escape HTML entities.
+# @category String
+def string.escape.html(s) =
+  string.replace(
+    pattern = "[&<>\"']",
+    fun (c) -> begin
+      if c == "&" then "&amp;"
+      elsif c == "<" then "&lt;"
+      elsif c == ">" then "&gt;"
+      elsif c == "\"" then "&quot;"
+      elsif c == "'" then "&apos;"
+      else c
+      end
+    end,
+    s
+  )
```
```suggestion
  r/[&<>"']/g.replace(
    fun (c) -> begin
      if c == "&" then "&amp;"
      elsif c == "<" then "&lt;"
      elsif c == ">" then "&gt;"
      elsif c == '"' then "&quot;"
      elsif c == "'" then "&#39;"
      else c
      end
    end,
    s
  )
```
I'd prefer to use [`r/_/g.replace(fn, str)`](https://www.liquidsoap.info/doc-dev/language.html#regular-expressions) regular expression syntax.

ghostnumber7 replied:
the `&#` aproach is actually shorter to implement if I can get the char code in any way (and would use it for all of them in that case), but could not find a way to do that. Is there any?

vitoyucepi commented:
Could you add some [tests](https://github.com/savonet/liquidsoap/blob/main/tests/language/string.liq#L115)?

```ruby
def test_escape_html() =
  test.equals(string.escape.html("&"), "&amp;")
  test.equals(string.escape.html("<"), "&lt;")
  test.equals(string.escape.html(">"), "&gt;")
  test.equals(string.escape.html('"'), "&quot;")
  test.equals(string.escape.html("'"), "&#39;")
  test.equals(string.escape.html("&<>\"'"), "&amp;&lt;&gt;&quot;&#39;")
  test.equals(string.escape.html("not escaped"), "not escaped")
  test.equals(string.escape.html("&quot;double escape&quot;"), "&amp;quot;double escape&amp;quot;")
  test.equals(string.escape.html("\\"), "\\")
  test.equals(string.escape.html("/"), "/")
  test.equals(string.escape.html("`"), "`")
end

test.check(test_escape_html)
```

ghostnumber7 commented:
> Could you add some [tests](https://github.com/savonet/liquidsoap/blob/main/tests/language/string.liq#L115)?
> 
> ```ruby
> def test_escape_html() =
>   test.equals(string.escape.html("&"), "&amp;")
>   test.equals(string.escape.html("<"), "&lt;")
>   test.equals(string.escape.html(">"), "&gt;")
>   test.equals(string.escape.html('"'), "&quot;")
>   test.equals(string.escape.html("'"), "&#39;")
>   test.equals(string.escape.html("&<>\"'"), "&amp;&lt;&gt;&quot;&#39;")
>   test.equals(string.escape.html("not escaped"), "not escaped")
>   test.equals(string.escape.html("&quot;double escape&quot;"), "&amp;quot;double escape&amp;quot;")
>   test.equals(string.escape.html("\\"), "\\")
>   test.equals(string.escape.html("/"), "/")
>   test.equals(string.escape.html("`"), "`")
> end
> 
> test.check(test_escape_html)
> ```

might be a dumb question, but how do I run tests locally to see they are working? :)

ghostnumber7 commented:
@vitoyucepi how about doing this instead of the current approach? It should work just the same 

```
# Escape HTML entities.
# @category String
def string.escape.html(s) =
   r/[&<>"']/g.replace(
    fun (c) -> "&##{string.binary.to_int(c)};",
    s
  )
end
```

vitoyucepi commented:
> might be a dumb question, but how do I run tests locally to see they are working? :)

Sorry, I haven't yet figured out how to build liquidsoap from the local source using dune.
I pretend everything works fine and wait for the CI results.
